File: arrays/search_a_2d_matrix.rb
def search_matrix(matrix, target)
    # flatten the matrix
    flat = []
    matrix.each do |row|
        row.each do |num|
            flat << num
        end
    end

    # set right and left
    left = 0
    right = flat.size - 1

    while left <= right
        return false if flat.include?(target) == false
        mid = (left + right) / 2
        if flat[mid] == target 
            return flat[mid] == target
        elsif flat[mid] < target
            left = mid + 1
        else
            right = mid - 1
        end
    end

end

File: arrays/search_a_2d_matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # flatten the matrix
        flat = []

        for row in range(len(matrix)):
            for num in range(row):
                flat.append(num)

        # set right and left
        left = 0
        right = len(flat) - 1

        while left <= right:
            if target not in flat:
                return False

            mid = (left + right) / 2
            if flat[mid] == target: 
                return flat[mid] == target
            elif flat[mid] < target:
                left = mid + 1
            else:
                right = mid - 1